### ISSUE TITLE
RuntimeWarning: networkx backend defined more than once for networkx v3.2+ on RHEL 9.2 OS

### ISSUE THREAD

**Issue by matthewfeickert** [networkx/networkx#7101]
This isn't specifically a "bug" in that anything is broken, but more a curiosity (to me) that a RuntimeWarning seems to be OS dependent. For installation in RHEL 9.2 releases I'm seeing

```pytb
RuntimeWarning: networkx backend defined more than once: nx-loopback
  backends.update(_get_backends("networkx.backends"))
```

but this isn't present on other OS versions. I'm not sure why.

### Current Behavior

To get a REHL 9.2 environment (which is what the cluster I'm on has) we can use Docker or podman and RedHat's container registries 

```console
$ docker pull registry.access.redhat.com/ubi9/ubi:9.2-755.1697625012
$ docker run --rm -ti registry.access.redhat.com/ubi9/ubi:9.2-755.1697625012
[root@6c881a6cd600 /]# cat /etc/os-release 
NAME="Red Hat Enterprise Linux"
VERSION="9.2 (Plow)"
ID="rhel"
ID_LIKE="fedora"
VERSION_ID="9.2"
PLATFORM_ID="platform:el9"
PRETTY_NAME="Red Hat Enterprise Linux 9.2 (Plow)"
ANSI_COLOR="0;31"
LOGO="fedora-logo-icon"
CPE_NAME="cpe:/o:redhat:enterprise_linux:9::baseos"
HOME_URL="https://www.redhat.com/"
DOCUMENTATION_URL="https://access.redhat.com/documentation/en-us/red_hat_enterprise_linux/9"
BUG_REPORT_URL="https://bugzilla.redhat.com/"

REDHAT_BUGZILLA_PRODUCT="Red Hat Enterprise Linux 9"
REDHAT_BUGZILLA_PRODUCT_VERSION=9.2
REDHAT_SUPPORT_PRODUCT="Red Hat Enterprise Linux"
REDHAT_SUPPORT_PRODUCT_VERSION="9.2"
[root@6c881a6cd600 /]# python3 --version
Python 3.9.16
[root@6c881a6cd600 /]# dnf install -y python3-pip
[root@6c881a6cd600 /]# python3 -m venv venv && . venv/bin/activate && python3 -m pip --quiet install --upgrade pip setuptools wheel
(venv) [root@6c881a6cd600 /]# python -m pip --quiet install --upgrade 'networkx<3.2.1' && python -m pip freeze | grep networkx && python -c 'import networkx'
networkx==3.2
/venv/lib64/python3.9/site-packages/networkx/utils/backends.py:135: RuntimeWarning: networkx backend defined more than once: nx-loopback
  backends.update(_get_backends("networkx.backends")
(venv) [root@6c881a6cd600 /]# python -m pip --quiet install --upgrade 'networkx<3.2' && python -m pip freeze | grep networkx && python -c 'import networkx'
networkx==3.1
(venv) [root@6c881a6cd600 /]# 
```

I don't really understand why this 

```pytb
RuntimeWarning: networkx backend defined more than once: nx-loopback
  backends.update(_get_backends("networkx.backends")
```

pops up for `networkx` `v3.2+`. This doesn't happen in a Debian based container:

```console
$ docker run --rm -ti python:3.9 /bin/bash
root@5eb063d4ac32:/# python -m venv venv && . venv/bin/activate
(venv) root@5eb063d4ac32:/# python3 -m pip --quiet install --upgrade pip setuptools wheel
(venv) root@5eb063d4ac32:/# python -m pip --quiet install --upgrade 'networkx<3.2.1' && python -m pip freeze | grep networkx && python -c 'import networkx'
networkx==3.2
(venv) root@5eb063d4ac32:/#
```

### Expected Behavior

No `RuntimeWarning` is raised.

### Steps to Reproduce

```console
$ docker run --rm -ti registry.access.redhat.com/ubi9/ubi:9.2-755.1697625012
# dnf install -y python3-pip
# python3 -m venv venv && . venv/bin/activate && python3 -m pip --quiet install --upgrade pip setuptools wheel
# python -m pip --quiet install --upgrade 'networkx<3.2.1' && python -m pip freeze | grep networkx && python -c 'import networkx'
```

### Environment

Python version: Python 3.9.16
NetworkX version: 3.2+

**Comment by MridulS** [comment]
Thanks for the detailed report @matthewfeickert!

Yeah this is weird. For RHEL the `nx-loopback` entry point is registered twice during the pip install. Inside the RHEL container:
``` python
>>> from importlib.metadata import entry_points
>>> entry_points()["networkx.backends"]
(EntryPoint(name='nx-loopback', value='networkx.classes.tests.dispatch_interface:dispatcher', group='networkx.backends'),
EntryPoint(name='nx-loopback', value='networkx.classes.tests.dispatch_interface:dispatcher', group='networkx.backends'))
```

For the debian one it register the entry point only once as expected:
```python
>>> from importlib.metadata import entry_points
>>> entry_points()["networkx.backends"]
(EntryPoint(name='nx-loopback', value='networkx.classes.tests.dispatch_interface:dispatcher', group='networkx.backends'),)
```

Need to investigate this more.

**Comment by MridulS** [comment]
Just circling back on this, I'm no RHEL expert but I can't reproduce this with python3.11 in the same container btw. (`python3.9.16` issue still persists)
```
$ dnf install -y python3.11-pip
$ python3.11 -m venv venv && . venv/bin/activate && python3.11 -m pip --quiet install --upgrade pip setuptools wheel
$ python -m pip --quiet install --upgrade 'networkx' && python -m pip freeze | grep networkx && python -c 'import networkx'
networkx==3.2.1
```

and pip registers the entry point properly (only once)

``` python
>>> from importlib.metadata import entry_points
>>> entry_points()["networkx.backends"]
<stdin>:1: DeprecationWarning: SelectableGroups dict interface is deprecated. Use select.
[EntryPoint(name='nx-loopback', value='networkx.classes.tests.dispatch_interface:dispatcher', group='networkx.backends')]
```

Is it possible the packaged python3.9/pip in RHEL 9.2 is buggy itself?

**Comment by matthewfeickert** [comment]
> Is it possible the packaged python3.9/pip in RHEL 9.2 is buggy itself?

I don't know, but possbile.

I haven't spent any additional time debugging this, but given what you've shown I'm happy to consider this out of scope / not `networkx`'s bug to fix.

**Comment by rossbar** [comment]
Thanks @matthewfeickert and @MridulS for investigating. I'll go ahead and close this, but if there's any new info on what's happening please feel free to re-open or if there's anything NetworkX needs to do to better support entrypoints on other systems, feel free to open new issues/PRs!

**Comment by notatallshaw-gts** [comment]
FYI, for anyone who wants to suppress the warning (pollutes some CI pipelines I run), you can do the following:

```python
import warnings

with warnings.catch_warnings():
    warnings.filterwarnings(
        "ignore", message="networkx backend defined more than once: nx-loopback"
    )
    import networkx as nx
```

**Comment by ArthurGarnier** [comment]
Hi,
Redhat will not fix this. 
https://issues.redhat.com/browse/RHEL-26149

Regards,

**Comment by MridulS** [comment]
> Redhat will not fix this.
> https://issues.redhat.com/browse/RHEL-26149

Not sure what we can do here, py3.9 will be dropped soon in networkx in the next release and seems like there is a quick solution here https://github.com/networkx/networkx/issues/7101#issuecomment-1919737867 if you do plan on using the old versions :)
